modules/dreambooth/dreambooth.py

```diff
@@ -0,0 +1,789 @@
+                            noise, noise_prior = torch.chunk(noise, 2, dim=0)
+
+                            # Compute instance loss
+                            loss = F.mse_loss(noise_pred.float(), noise.float(), reduction="none").mean(
+                                [1, 2, 3]).mean()
+
+                            # Compute prior loss
+                            prior_loss = F.mse_loss(noise_pred_prior.float(), noise_prior.float(), reduction="mean")
+
+                            # Add the prior loss to the instance loss.
+                            loss = loss + self.prior_loss_weight * prior_loss
+                        else:
+                            loss = F.mse_loss(noise_pred.float(), noise.float(), reduction="mean")
+
+                        accelerator.backward(loss)
+                        if accelerator.sync_gradients:
+                            params_to_clip = (
+                                itertools.chain(unet.parameters(), text_encoder.parameters())
+                                if self.train_text_encoder
+                                else unet.parameters()
+                            )
+                            accelerator.clip_grad_norm_(params_to_clip, self.max_grad_norm)
+                        optimizer.step()
+                        lr_scheduler.step()
+                        optimizer.zero_grad()
+                        # Save checkpoint data *before* generating preview image
+                        if self.save_data_every:
+                            # Check to make sure this doesn't throw OOM if training on CPU
+                            if (not global_step % self.save_data_every and global_step != 0) or \
+                                    (not global_step % self.save_preview_every and global_step != 0):
+                                if accelerator.is_main_process:
+                                    print(f"Saving pretrained model data {global_step}.")
+                                    pipeline = StableDiffusionPipeline.from_pretrained(
+                                        self.pretrained_model_path,
+                                        unet=accelerator.unwrap_model(unet),
+                                        revision=self.total_steps + global_step
+                                    )
+                                    pipeline = pipeline.to("cuda")
+                                    with autocast("cuda"):
+                                        pipeline.text_encoder.resize_token_embeddings(49409)
+                                        if not global_step % self.save_data_every and global_step != 0:
+                                            try:
+                                                print(f"Saving to {self.output_dir}")
+                                                pipeline.save_pretrained(self.output_dir)
+                                                save_checkpoint(self.total_steps + global_step, self.src,
```

Comment:
IMO it's a good idea to allow `save_checkpoint` to receive a pipeline (instead of a path on disk) to save on I/O & RAM

Reply:
Good point. I can implement this, just need to decide how to handle the "save_checkpoint" call that runs in "start_training".

It's really not necessary if I update the logic a bit at the "check for save" part to ensure we're saving on the last iteration of training. Maybe I'm already doing this...